### ISSUE TITLE
Событие PaymentTypeChanged (V7Preview6)

### ISSUE THREAD

**Issue by igorpooh1978** [iiko/front.api.sdk#226]
Доброй ночи.

Подскажите, что за событие PaymentTypeChanged и когда оно срабатывает.
Судя по названию, оно должно срабатывать на экране проведения платежей.

Спасибо

**Comment by CLTanuki** [comment]
Добрый день.
Событие вызывается в тот момент, когда изменён тип оплаты (собственно, PaymentType). Например, изменён счёт проведения, либо тип отключён, либо иное.

**Comment by igorpooh1978** [comment]
Добрый вечер.
Понятно, что ничего не понятно
итак
я делаю подписку на данное событие
```
            subscriptions = new CompositeDisposable
            {
                PluginContext.Notifications.PaymentTypeChanged.Subscribe(PaymentChanged)

           };
```
и соответственно сам обработчик
```
        private void PaymentChanged(EntityChangedEventArgs<IPaymentType> obj)
        {
            throw new NotImplementedException();
        }
```
Если я на экране оплат добавлю или удалю платеж, то я должен попасть в PaymentChanged?

Если это так, то в V7Pr6 так не происходит и я не вижу входа в подписку

**Comment by CLTanuki** [comment]
Попадание в этот блок не зависит от экрана оплат и\или текущего заказа.
Для отслеживания изменений в заказе стоит использовать подписку OnOrderChanged.

**Comment by igorpooh1978** [comment]
Только что проверил
На удаление оплаты через Gui на экране платежа, OnOrderChanged - не реагирует

**Comment by CLTanuki** [comment]
Да. Это боль. Связана она с тем, что в модели заказа не появляется свежедобавленного платежа.
Обычно в таких случаях мы храним слепки заказа и сравниваем с ними изменения.
Также может помочь переход на другой экран - это уже особенности подписок.
Надеюсь, коллеги более подробно расскажут. А мы как-то уже давно свыклись и обходим эти углы.

**Comment by igorpooh1978** [comment]
Согласен.
В моём случае платежный экран - это последняя инстанция, на котором я должен
а. Или добавить товар со свободной ценой при добавлении платежа
б. Если удалить товар со свободной ценой в случае удаления платежа.
И если пункт А я добился, то пункт Б в случае возврата на экран редактированя заказа я удаляю товар со свободной ценой, но вот именно выполнить действие при удалении платежа на экране платежей - пока не очень понимаю что делать. И PaymentTypeChanged судя по названию д.б. подходить.
Подождем старших товарищей. Надеюсь они прояснят как такое сделать

**Comment by igorpooh1978** [comment]
Коллеги?

**Comment by vcpp** [comment]
Событие `PaymentTypeChanged` уведомляет об изменении типа оплаты (объекта типа `IPaymentType`). Типы оплаты могут меняться через iikoOffice (создание/удаление, изменение настроек и т.п.). По-хорошему, надо какой-то осмысленный комментарий к этому событию написать.
По изменениям внутри заказа отдельных событий нет, только общее `OrderChanged` и дальше сравнивать, что поменялось с прошлого раза.
На экране оплаты для скорости не все изменения сохраняются сразу, они могут накапливаться и сохраняться пачками, поэтому и плагин будет получать уведомления не сразу, а по мере сохранения изменений. Стоит отключать эту оптимизацию при наличии подписчиков `OrderChanged`?
Как насчёт добавления точки расширения перед оплатой? То есть кассир сделал всё, что хотел, нажимает "Оплатить", и в этот момент плагин может вклиниться и что-нибудь поменять напоследок (и даже отменить начало оплаты). Или надо чтоб кассир видел изменения именно при смене типа оплаты? Тогда можно добавить уведомление об изменении набора оплат.

**Comment by CLTanuki** [comment]
Логическая коллизия лишь в том, что при 0 сумме оплат платёж отменится сам. А иногда гость попросит отменить оплату, например, баллами. Так как плагин не видит нового платежа, его суммы и причины отмены, а только добавление, это приходилось угадывать. Впрочем, сейчас при появлении подписок на прокатывание карты и т.п. стало гораздо легче, так как для операций накопления используются другие сценарии. Впрочем, увы, не всегда это возможно.

**Comment by igorpooh1978** [comment]
Добрый день. 
Спасибо за реакцию )

> Стоит отключать эту оптимизацию при наличии подписчиков `OrderChanged`?
Ну при удалении оплаты с платежного экрана. Наверное, реакция всё же д.б., т.к. пример есть система ваучеров, в которой я могу провести оплаты только перед выполнением метода Pay в платежном плагине. Сейчас я как раз попал в ситуацию, когда я делаю оплаты в CollectData, далее добавляю в заказ доп. товар (для округления до суммы ваучеров) и вношу товар с открытой ценой на методе OnPaymentAdded
И вот клиент по недоразумению удаляет платеж ваучерами. До факта нажатия на кнопку Pay более об этих ваучерах и уж об их списании система не знает, да в платеже она никак не отразиться. А вот при попытке удаления платежа на платежном экране, по крайней мере можно сделать попытку возврата этих ваучеров обратно клиенту. Прошу прощения если сумбурно.

> То есть кассир сделал всё, что хотел, нажимает "Оплатить", и в этот момент плагин может вклиниться и что-нибудь поменять напоследок (и даже отменить начало оплаты)
Тоже не плохой вариант. Разблокировать заказ перед оплатой (аля BeforeDoCheque cделять BeforePay с разблокированным заказом). Придётся логики поломать, но тоже вариант. Но опять таки если платеж УЖЕ не участвует в оплате к плагину обращения не будет.
Сейчас же мне придётся подсаживаться на OrderChanged, чтобы при работе платежного плагина я по по окончанию понимал если в заказе до платежа были ваучера, а уже после оплаты тип оплаты ваучерами неведомым образом пропал, то в этом случае они должны отмениться.



> Или надо чтоб кассир видел изменения именно при смене типа оплаты? Тогда можно добавить уведомление об изменении набора оплат
Хочется всего и много. Но и для уведомления, что в моем случае ваучеры удалились кассир должен видеть

**Comment by igorpooh1978** [comment]
> Логическая коллизия лишь в том, что при 0 сумме оплат платёж отменится сам. А иногда гость попросит отменить оплату, например, баллами. Так как плагин не видит нового платежа, его суммы и причины отмены, а только добавление, это приходилось угадывать. Впрочем, сейчас при появлении подписок на прокатывание карты и т.п. стало гораздо легче, так как для операций накопления используются другие сценарии. Впрочем, увы, не всегда это возможно.

Это прекрасно, когда карта прокатывается. Плохо когда у тебя данные летят с пин-пада и с чипа(

**Comment by LeeTwelve** [comment]
> Как насчёт добавления точки расширения перед оплатой? То есть кассир сделал всё, что хотел, нажимает "Оплатить", и в этот момент плагин может вклиниться и что-нибудь поменять напоследок (и даже отменить начало оплаты). Или надо чтоб кассир видел изменения именно при смене типа оплаты? Тогда можно добавить уведомление об изменении набора оплат.

@vcpp Добрый день, а не подскажите как отменить начало оплаты?

**Comment by vcpp** [comment]
> Пытался бросить эксепшен в BeforeOrderBill, но оплата все равно прошла

Какой тип исключения бросали? Для отмены печати пречека из [BeforeOrderBill](https://iiko.github.io/front.api.sdk/v7/html/P_Resto_Front_Api_INotificationService_BeforeOrderBill.htm) надо бросать именно `OperationCanceledException`.

**Comment by LeeTwelve** [comment]
> > Пытался бросить эксепшен в BeforeOrderBill, но оплата все равно прошла
> 
> Какой тип исключения бросали? Для отмены печати пречека из [BeforeOrderBill](https://iiko.github.io/front.api.sdk/v7/html/P_Resto_Front_Api_INotificationService_BeforeOrderBill.htm) надо бросать именно `OperationCanceledException`.

Именно такой и использовал:
```c#
private void BeforeOrderBill(IOrder order, IUser user, IOperationService os, IViewManager vm)
 {
        Log.Info("Start BeforeOrderBill");
        throw new OperationCanceledException("OperationCanceledException");
}
```
Использую `Resto.Front.Api.V8Preview5 8.4.6018-alpha`
